### PR TITLE
avoid allocating 16 hashmaps per transaction

### PR DESCRIPTION
**Pull request by msmouse**
## Description

Instead, introduce TransactionsToKeep and let it carry a per shard state updates index (by refs).
StateKey-Values are still cloned unnecessarily but will clean further later.

Small performance improvement is expected.

## How Has This Been Tested?

existing coverage.

## Type of Change
- [x] New feature

## Which Components or Systems Does This Change Impact?
- [x] Validator Node

### CODE REVIEW

File: storage/storage-interface/src/sharded_state_update_refs.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    state_store::{state_key::StateKey, state_value::StateValue, NUM_STATE_SHARDS},
+    write_set::WriteSet,
+};
+use arr_macro::arr;
+
+pub type StateUpdateRefWithOffset<'kv> = (usize, &'kv StateKey, Option<&'kv StateValue>);
+
+pub struct ShardedStateUpdateRefs<'kv> {
+    /// (idx, key, value) tuple per update, sharded
+    ///    Box<[]> to release over allocated memory during construction
```
Comment (grao1991) [inline storage/storage-interface/src/sharded_state_update_refs.rs:14]: what does this comment mean?

File: storage/storage-interface/src/sharded_state_update_refs.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    pub fn index_per_version_updates<
+        UpdateIter: IntoIterator<Item = (&'kv StateKey, Option<&'kv StateValue>)>,
+        VersionIter: IntoIterator<Item = UpdateIter>,
+    >(
+        updates_by_version: VersionIter,
+        num_versions: usize,
+    ) -> Self {
+        let mut shards = arr![Vec::with_capacity(num_versions / 16); 16];
```
Comment (grao1991) [inline storage/storage-interface/src/sharded_state_update_refs.rs:39]: why / 16?

File: storage/storage-interface/src/sharded_state_update_refs.rs (hunk 1 above)
Comment (msmouse) [reply storage/storage-interface/src/sharded_state_update_refs.rs:14]: when `vec` converts to `box<[]>`, over allocated memory is freezed.

File: storage/storage-interface/src/sharded_state_update_refs.rs (hunk 2 above)
Comment (msmouse) [reply storage/storage-interface/src/sharded_state_update_refs.rs:39]: don't want to over allocate too much. As long as it's not empty, we shouldn't hit too many reallocates, right?

Comment (grao1991) [reply storage/storage-interface/src/sharded_state_update_refs.rs:43]: I'd say at least `/ 8` since most of txns touches at least 2 states.

**Review by grao1991 — APPROVED** [review]

**Review by areshand — APPROVED** [review]

File: types/src/state_store/mod.rs [hunk 3]
```diff
@@ -129,22 +132,53 @@ impl<K: Clone + Eq + Hash> TStateView for MockStateView<K> {
...
-        lhs.par_iter_mut()
-            .zip_eq(rhs.par_iter())
+    pub fn merge(&mut self, other: Self) {
+        self.shards
+            .par_iter_mut()
+            .zip_eq(other.shards.into_par_iter())
             .for_each(|(l, r)| {
-                l.extend(r.clone());
+                l.extend(r);
             })
```
Comment (msmouse) [inline types/src/state_store/mod.rs:161]: this used to be in the "exe pool", now it's in the global pool

**Pull request merged**